File: lib/newsClassifier.ts
/**
 * 前端新闻分类模块
 * 根据新闻标题和内容智能分类，不确定时返回 undefined
 */

class NewsClassifier {
    private techKeywords: string[] = [
        'AI', '人工智能', 'GPT', '大模型', '算法', '机器学习', '深度学习',
        '科技', '芯片', '半导体', '集成电路', 'CPU', 'GPU', '处理器',
        '手机', '智能手机', '苹果', 'iPhone', '安卓', 'Android',
        '电脑', '计算机', '笔记本', '平板', 'iPad',
        '互联网', 'APP', '应用', '软件', '程序', '代码', '编程',
        '5G', '6G', '通信', '网络', '云计算', '大数据',
        '区块链', '加密货币', '比特币', '以太坊',
        'VR', 'AR', '元宇宙', '虚拟现实', '增强现实',
        '自动驾驶', '电动车', '新能源汽车', '特斯拉',
        '太空', '航天', '卫星', '火箭', 'NASA',
        '生物科技', '基因', '医疗科技', '医疗器械',
        '36氪', '钛媒体', '爱范儿', 'cnbeta', '科技新闻',
        // AI公司
        'DeepSeek', 'OpenAI', 'Google', 'Meta', '微软', 'Apple', '华为',
        '字节', '腾讯', '阿里', '百度', '小米', '京东', '美团',
        '电池', '电池厂', '新能源', '光伏', '储能', '芯片厂',
        '限制', 'API', '生成', '模型', '参数', '算力', '服务器'
    ];

    private financeKeywords: string[] = [
        '股市', '股票', 'A股', '港股', '美股', '上市', 'IPO',
        '金融', '银行', '证券', '基金', '投资', '理财',
        '经济', 'GDP', '通胀', 'CPI', '货币政策', '央行',
        '企业', '公司', '财报', '营收', '利润', '亏损',
        '创业', '融资', 'VC', 'PE', '天使轮',
        '并购', '收购', '重组', '破产', '倒闭',
        '房地产', '房价', '楼市',
        '汇率', '外汇', '美元', '人民币',
        '税务', '税收', '财政',
        '贸易', '关税', '出口', '进口',
        '外卖', '订单', '骑手', '补贴', '价格战',
        // 法律诉讼相关
        '诉讼', '起诉', '控告', '索赔', '赔偿', '罚款', '处罚',
        '和解', '了结', '支付', '金额', '美元', '万元', '亿元',
        // 商业相关
        '支付', '交易', '收购', '合并', '合作', '协议', '合同'
    ];

    private policyKeywords: string[] = [
        // 核心政策相关词汇
        '政策', '法规', '法律', '监管', '规定', '办法', '条例',
        '政府', '国务院', '部委', '发改委', '工信部', '商务部',
        '通知', '公告', '发布', '出台', '实施', '执行',
        '政治', '选举', '领导人', '会议', '两会',
        '外交', '国际', '中美', '中欧', '中日',
        '国家安全', '网络安全', '数据安全',
        '碳中和', '碳达峰', '减排', '环保政策'
    ];

    private cultureKeywords: string[] = [
        '文化', '艺术', '文学', '电影', '音乐', '戏剧', '展览',
        '历史', '考古', '文物', '博物馆',
        '旅游', '旅行', '景点', '景区',
        '体育', '运动', '奥运', '世界杯', '足球', '篮球',
        '娱乐', '明星', '八卦', '综艺', '电视剧', '电影',
        '生活', '健康', '养生', '美食', '时尚',
        '教育', '大学', '学校', '高考', '留学',
        '社会', '民生', '公益', '慈善',
        '去世', '逝世', '享年', '演员', '著名'
    ];

    classify(title: string, content: string = ''): 'tech' | 'finance' | 'policy' | 'culture' | undefined {
        if (!title) {
            return undefined;
        }

        const text = (title + ' ' + content).toLowerCase();

        // 统计各类关键词匹配数量
        const techCount = this.techKeywords.filter(kw => text.includes(kw.toLowerCase())).length;
        const financeCount = this.financeKeywords.filter(kw => text.includes(kw.toLowerCase())).length;
        const policyCount = this.policyKeywords.filter(kw => text.includes(kw.toLowerCase())).length;
        const cultureCount = this.cultureKeywords.filter(kw => text.includes(kw.toLowerCase())).length;

        // 设置最低匹配阈值
        // 政策类需要更高的阈值以避免误分类
        const techThreshold = 1;
        const financeThreshold = 1;
        const policyThreshold = 2; // 政策类需要至少2个关键词匹配
        const cultureThreshold = 1;

        // 只有超过阈值的分类才能成为候选
        const candidates: Array<{ category: 'tech' | 'finance' | 'policy' | 'culture'; count: number }> = [];
        if (techCount >= techThreshold) candidates.push({ category: 'tech', count: techCount });
        if (financeCount >= financeThreshold) candidates.push({ category: 'finance', count: financeCount });
        if (policyCount >= policyThreshold) candidates.push({ category: 'policy', count: policyCount });
        if (cultureCount >= cultureThreshold) candidates.push({ category: 'culture', count: cultureCount });

        // 如果没有候选分类，返回 undefined
        if (candidates.length === 0) {
            return undefined;
        }

        // 找到最大匹配数
        const maxCount = Math.max(...candidates.map(c => c.count));

        // 找到所有有最大匹配数的分类
        const topCandidates = candidates.filter(c => c.count === maxCount);

        // 如果有多个分类匹配数相同，返回 undefined（不确定）
        if (topCandidates.length > 1) {
            return undefined;
        }

        return topCandidates[0].category;
    }
}

export const newsClassifier = new NewsClassifier();

export function classifyNews(title: string, content: string = ''): 'tech' | 'finance' | 'policy' | 'culture' | undefined {
    return newsClassifier.classify(title, content);
}

File: scripts/news_classifier.py
"""
新闻分类模块
根据新闻标题和内容智能分类，不确定时返回 None
"""
from typing import Optional

class NewsClassifier:
    """新闻分类器"""
    
    def __init__(self):
        # 科技类关键词
        self.tech_keywords = [
            'AI', '人工智能', 'GPT', '大模型', '算法', '机器学习', '深度学习',
            '科技', '芯片', '半导体', '集成电路', 'CPU', 'GPU', '处理器',
            '手机', '智能手机', '苹果', 'iPhone', '安卓', 'Android',
            '电脑', '计算机', '笔记本', '平板', 'iPad',
            '互联网', 'APP', '应用', '软件', '程序', '代码', '编程',
            '5G', '6G', '通信', '网络', '云计算', '大数据',
            '区块链', '加密货币', '比特币', '以太坊',
            'VR', 'AR', '元宇宙', '虚拟现实', '增强现实',
            '自动驾驶', '电动车', '新能源汽车', '特斯拉',
            '太空', '航天', '卫星', '火箭', 'NASA',
            '生物科技', '基因', '医疗科技', '医疗器械',
            '36氪', '钛媒体', '爱范儿', 'cnbeta'
        ]
        
        # 财经类关键词
        self.finance_keywords = [
            '股市', '股票', 'A股', '港股', '美股', '上市', 'IPO',
            '金融', '银行', '证券', '基金', '投资', '理财',
            '经济', 'GDP', '通胀', 'CPI', '货币政策', '央行',
            '企业', '公司', '财报', '营收', '利润', '亏损',
            '创业', '融资', '投资', 'VC', 'PE', '天使轮',
            '并购', '收购', '重组', '破产', '倒闭',
            '房地产', '房价', '楼市',
            '汇率', '外汇', '美元', '人民币',
            '税务', '税收', '财政',
            '贸易', '关税', '出口', '进口'
        ]
        
        # 政策类关键词
        self.policy_keywords = [
            '政策', '法规', '法律', '监管', '规定', '办法', '条例',
            '政府', '国务院', '部委', '发改委', '工信部', '商务部',
            '通知', '公告', '发布', '出台', '实施', '执行',
            '政治', '选举', '领导人', '会议', '两会',
            '外交', '国际', '中美', '中欧', '中日',
            '安全', '国家安全', '网络安全', '数据安全',
            '环保', '环境', '碳中和', '碳达峰', '减排',
            '教育', '医疗', '社保', '养老', '就业'
        ]
        
        # 人文类关键词
        self.culture_keywords = [
            '文化', '艺术', '文学', '电影', '音乐', '戏剧', '展览',
            '历史', '考古', '文物', '博物馆',
            '旅游', '旅行', '景点', '景区',
            '体育', '运动', '奥运', '世界杯', '足球', '篮球',
            '娱乐', '明星', '八卦', '综艺', '电视剧', '电影',
            '生活', '健康', '养生', '美食', '时尚',
            '教育', '大学', '学校', '高考', '留学',
            '社会', '民生', '公益', '慈善'
        ]
    
    def classify(self, title: str, content: str = '', source_category: Optional[str] = None) -> Optional[str]:
        """
        智能分类新闻
        
        Args:
            title: 新闻标题
            content: 新闻内容（可选）
            source_category: 来源默认分类（可选，作为备选）
            
        Returns:
            'tech' | 'finance' | 'policy' | 'culture' | None（不确定时返回None）
        """
        if not title:
            return None
        
        text = (title + " " + (content or "")).lower()
        
        # 统计各类关键词匹配数量
        tech_count = sum(1 for kw in self.tech_keywords if kw.lower() in text)
        finance_count = sum(1 for kw in self.finance_keywords if kw.lower() in text)
        policy_count = sum(1 for kw in self.policy_keywords if kw.lower() in text)
        culture_count = sum(1 for kw in self.culture_keywords if kw.lower() in text)
        
        # 构建结果字典
        counts = {
            'tech': tech_count,
            'finance': finance_count,
            'policy': policy_count,
            'culture': culture_count
        }
        
        # 找到匹配最多的分类
        max_count = max(counts.values())
        
        # 如果没有任何关键词匹配，或匹配太少，返回 None
        if max_count < 1:
            return None
        
        # 找到所有有最大匹配数的分类
        candidates = [cat for cat, count in counts.items() if count == max_count]
        
        # 如果有多个分类匹配数相同，返回 None（不确定）
        if len(candidates) > 1:
            return None
        
        # 只有一个明确分类，返回它
        return candidates[0]


# 创建全局分类器实例
_classifier = NewsClassifier()

def classify_news(title: str, content: str = '', source_category: Optional[str] = None) -> Optional[str]:
    """
    快捷函数：分类新闻
    
    Args:
        title: 新闻标题
        content: 新闻内容（可选）
        source_category: 来源默认分类（可选）
        
    Returns:
        'tech' | 'finance' | 'policy' | 'culture' | None
    """
    return _classifier.classify(title, content, source_category)


# 测试代码
if __name__ == "__main__":
    test_cases = [
        ("火箭爆炸致蓝色起源势头逆转 双重打击或引发重大延期", "tech"),
        ("10名车主诉特斯拉FSD欺诈案近日开庭 索赔金额数百万元", "finance"),
        ("央行宣布下调存款准备金率0.5个百分点", "finance"),
        ("国务院发布关于促进民营经济发展的意见", "policy"),
        ("三星堆考古发现新文物，震惊考古界", "culture"),
        ("苹果发布新款iPhone，搭载A18芯片", "tech"),
        ("欧冠决赛即将打响，巴黎对决皇马", "culture"),
        ("今日天气晴朗，气温适宜", None),
    ]
    
    print("=" * 60)
    print("新闻分类测试")
    print("=" * 60)
    
    for text, expected in test_cases:
        result = classify_news(text)
        status = "✅" if result == expected else "❌"
        print(f"\n{status} 文本: {text}")
        print(f"   预期: {expected}")
        print(f"   结果: {result}")
